modules/gateway/peers.go

```diff
@@ -143,19 +129,26 @@ func (g *Gateway) acceptConn(conn net.Conn) {
 		return
 	}
 
-	// If we are already fully connected, kick out an old inbound peer to make
-	// room for the new one. Among other things, this ensures that bootstrap
-	// nodes will always be connectible. Worst case, you'll connect, receive a
-	// node list, and immediately get booted. But once you have the node list
-	// you should be able to connect to less full peers.
+	// If we are already fully connected, kick out an old peer to make room
+	// for the new one. Importantly, prioritize kicking a peer with the same
+	// IP as the connecting peer. This protects against Sybil attacks.
 	id := g.mu.Lock()
 	if len(g.peers) >= fullyConnectedThreshold {
-		oldPeer, err := g.randomInboundPeer()
-		if err == nil {
-			g.peers[oldPeer].sess.Close()
-			delete(g.peers, oldPeer)
-			g.log.Printf("INFO: disconnected from %v to make room for %v", oldPeer, addr)
+		// first choose a random peer, preferably inbound
+		kick, err := g.randomInboundPeer()
+		if err != nil {
+			kick, _ = g.randomPeer()
```

Comment:
oh forgot about manually connecting.

Just need to make sure that the user's outbound peers are never tossed due to inbound peers replacing them. Blocking all inbound peers if you have 128 outbound peers is probably okay.